File: backend/src/main/java/com/dreampath/service/rag/RagEmbeddingService.java
package com.dreampath.service.rag;

import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import okhttp3.*;
import org.json.JSONArray;
import org.json.JSONObject;

@Service
@RequiredArgsConstructor
public class RagEmbeddingService {

    @Value("${openai.api.key}")
    private String openaiApiKey;

    private final OkHttpClient client = new OkHttpClient();

    public float[] embed(String text) {

        try {
            JSONObject json = new JSONObject();
            json.put("model", "text-embedding-3-small");
            json.put("input", text);

            RequestBody body = RequestBody.create(
                    json.toString(),
                    MediaType.parse("application/json")
            );

            Request request = new Request.Builder()
                    .url("https://api.openai.com/v1/embeddings")
                    .addHeader("Authorization", "Bearer " + openaiApiKey)
                    .post(body)
                    .build();

            Response response = client.newCall(request).execute();
            String responseBody = response.body().string();

            JSONObject obj = new JSONObject(responseBody);
            var arr = obj.getJSONArray("data")
                    .getJSONObject(0)
                    .getJSONArray("embedding");

            float[] vector = new float[arr.length()];
            for (int i = 0; i < arr.length(); i++) {
                vector[i] = (float) arr.getDouble(i);
            }

            return vector;

        } catch (Exception e) {
            throw new RuntimeException("임베딩 생성 실패", e);
        }
    }
}

File: ai-service/services/chatbot/rag_embedding_service.py
import os
from openai import OpenAI
from typing import List


class RagEmbeddingService:
    def __init__(self):
        self.client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

    def embed(self, text: str) -> List[float]:
        """텍스트를 임베딩 벡터로 변환"""
        try:
            response = self.client.embeddings.create(
                model="text-embedding-3-small",
                input=text
            )
            return response.data[0].embedding
        except Exception as e:
            raise RuntimeError(f"임베딩 생성 실패: {str(e)}")
